Counterexamples/HeawoodUnitDistance.lean

```diff
@@ -0,0 +1,212 @@
+  revert i; decide
+
+/-! ### A key number -/
+
+namespace Heawood
+
+/-- `2c^3+3c+1` has a root in `(-1/3, -5/16)`. This root, which turns out to be unique
+(we do not need uniqueness), underpins our unit-distance embedding's coordinates. -/
+lemma exists_root_in_Ioo :
+    ∃ c ∈ Set.Ioo (-1 / 3 : ℝ) (-5 / 16), 2 * c ^ 3 + 3 * c + 1 = 0 :=
+  intermediate_value_Ioo (by norm_num) (by fun_prop) (by norm_num)
+
+/-- The real root `c = -0.312908...` of `2c^3+3c+1`. -/
+noncomputable def root : ℝ :=
+  exists_root_in_Ioo.choose
+
+@[inherit_doc] scoped notation "c" => root
+
+lemma root_bounds : c ∈ Set.Ioo (-1 / 3) (-5 / 16) :=
+  exists_root_in_Ioo.choose_spec.1
+
+lemma root_equation : 2 * c ^ 3 + 3 * c + 1 = 0 :=
+  exists_root_in_Ioo.choose_spec.2
+
+/-! ### The embedding proper -/
+
+/-- Notation for the Euclidean plane. -/
+notation "Plane" => EuclideanSpace ℝ (Fin 2)
+
+lemma dist_eq_one_iff {x₀ y₀ x₁ y₁ : ℝ} :
+    dist !₂[x₀, y₀] !₂[x₁, y₁] = 1 ↔ (x₀ - x₁) ^ 2 + (y₀ - y₁) ^ 2 = 1 := by
+  simp [dist_eq_norm_sub, PiLp.norm_eq_of_L2]
+
+/-- The base function from graph vertices to Euclidean points in our embedding. -/
+noncomputable def udMap : Fin 14 → Plane
+  | 1 => !₂[(1 + c) / 2, c ^ 2 - c / 2 + 1]
+  | 0 => !₂[c, 1 / 2] | 7 => !₂[0, 1 / 2] | 2 => !₂[1, 1 / 2] | 9 => !₂[1 - c, 1 / 2]
+  | 10 => !₂[(1 + c) / 2, c ^ 2 - c / 2] | 5 => !₂[(1 - c) / 2, c ^ 2 - c / 2]
+  | 3 => !₂[(1 + c) / 2, -(c ^ 2 - c / 2)] | 8 => !₂[(1 - c) / 2, -(c ^ 2 - c / 2)]
+  | 13 => !₂[c, -1 / 2] | 6 => !₂[0, -1 / 2] | 11 => !₂[1, -1 / 2] | 4 => !₂[1 - c, -1 / 2]
+  | 12 => !₂[(1 + c) / 2, -(c ^ 2 - c / 2 + 1)]
+
+lemma reflect_toEuclideanLin {x y : ℝ} : !![1, 0; 0, -1].toEuclideanLin !₂[x, y] = !₂[x, -y] := by
+  ext i; fin_cases i <;> simp
+
+lemma udMap_reflect (i : Fin 14) : udMap i.rev = !![1, 0; 0, -1].toEuclideanLin (udMap i) := by
+  fin_cases i <;> simp only [udMap, reflect_toEuclideanLin, Fin.reduceFinMk, Fin.reduceRev] <;>
+  norm_num
+
+lemma decompose_point (p : Plane) : p = !₂[p.proj 0, p.proj 1] := by
+  ext i; fin_cases i <;> simp
+
+/-- `udMap` is injective on indices `[0, 7, 10, 5, 2, 9]` because their x-coordinates
+strictly increase in that order. -/
+lemma injOn_udMap_sextet : Set.InjOn udMap ({0, 7, 10, 5, 2, 9} : Finset (Fin 14)) := by
+  let f : Fin 6 → Fin 14 := fun | 0 => 0 | 1 => 7 | 2 => 10 | 3 => 5 | 4 => 2 | 5 => 9
+  suffices StrictMono fun n ↦ (udMap (f n)).proj 0 by
+    intro i mi j mj h
+    rw [mem_coe] at mi mj
+    obtain ⟨ci, hci⟩ : ∃ ci, f ci = i := by fin_cases mi <;> decide
+    obtain ⟨cj, hcj⟩ : ∃ cj, f cj = j := by fin_cases mj <;> decide
+    rw [← hci, ← hcj] at h
+    apply_fun (·.proj 0) at h
+    rwa [← hci, this.injective h]
+  rw [Fin.strictMono_iff_lt_succ]
+  obtain ⟨lb, ub⟩ := root_bounds
+  intro k
+  fin_cases k
+  all_goals
+    simp only [f, udMap, Fin.reduceFinMk, Fin.reduceSucc, Fin.reduceCastSucc, PiLp.proj_apply,
+      Matrix.cons_val_zero]
+    linarith
```

Comment:
```suggestion
  simp [Fin.forall_fin_succ, f, udMap]
  grind
```
on a more recent version of Lean, this proof works, but perhaps it doesn't on this version... :(